enhancements/proposals/1978-scylladbdatacenter-api/README.md

```diff
@@ -0,0 +1,761 @@
+	// +optional
+	Annotations map[string]string `json:"annotations,omitempty"`
+}
+
+type RackTemplate struct {
+	// nodes specifies desired number of nodes in rack.
+	// +optional
+	Nodes *int32 `json:"nodes"`
+
+	// placement describes restrictions for the nodes ScyllaDB is scheduled on.
+	// +optional
+	Placement *Placement `json:"placement,omitempty"`
+
+	// scyllaDB provides overrides for properties of ScyllaDB for this rack.
+	// They override the settings set on Datacenter level.
+	ScyllaDB *ScyllaDBOverrides `json:"scyllaDB"`
+
+	// scyllaDBManagerAgent provides overrides for properties of ScyllaDB Manager Agent for this rack.
+	// They override the settings set on Datacenter level.
+	ScyllaDBManagerAgent *ScyllaDBManagerAgentOverrides `json:"scyllaDBManagerAgent"`
+
+	// topologyLabelSelector specifies a label selector which will be used to target nodes at specified topology constraints.
+	// Datacenter topologyLabelSelector is merged with rack topologyLabelSelector and then converted into nodeAffinity
+	// targeting nodes having specified topology.
+	TopologyLabelSelector *metav1.LabelSelector `json:"topologyLabelSelector,omitempty"`
+}
+
+// ScyllaDBOverrides allows overriding a subset of ScyllaDB settings.
+type ScyllaDBOverrides struct {
+	// resources requirements for the ScyllaDB container
+	// +optional
+	Resources *corev1.ResourceRequirements `json:"resources,omitempty"`
+
+	// storage requirements for the containers
+	// +optional
+	Storage *Storage `json:"storage,omitempty"`
+
+	// customConfigMapRef points to custom ScyllaDB configuration stored as ConfigMap.
+	// Overrides upper level settings.
+	// +optional
+	CustomConfigMapRef *string `json:"customConfigMapRef,omitempty"`
+
+	// volumes added to Scylla Pod.
+	// +optional
+	Volumes []corev1.Volume `json:"volumes,omitempty"`
+
+	// volumeMounts to be added to ScyllaDB container.
+	VolumeMounts []corev1.VolumeMount `json:"volumeMounts,omitempty"`
+}
+
+// ScyllaDBManagerAgentOverrides allows to override a subset of ScyllaManagerAgent settings.
+type ScyllaDBManagerAgentOverrides struct {
+	// resources are requirements for the ScyllaDB Manager Agent container.
+	// +optional
+	Resources *corev1.ResourceRequirements `json:"resources,omitempty"`
+
+	// customConfigSecretRef points to custom ScyllaDB Manager Agent configuration stored as Secret.
+	// +optional
+	CustomConfigSecretRef *string `json:"customConfigSecretRef,omitempty"`
+
+	// volumes added to Scylla Pod.
+	// +optional
+	Volumes []corev1.Volume `json:"volumes,omitempty"`
+
+	// volumeMounts to be added to Agent container.
+	VolumeMounts []corev1.VolumeMount `json:"volumeMounts,omitempty"`
+}
+
+// RackSpec is the desired state for a ScyllaDB Rack.
+type RackSpec struct {
+	RackTemplate
+
+	// name is the name of the ScyllaDB Rack. Used as rack name in GossipingPropertyFileSnitch.
+	Name string `json:"name"`
+}
+
+// ScyllaDB holds configuration options related to ScyllaDB.
+type ScyllaDB struct {
+	ScyllaDBOverrides
```

Comment:
I'd merged the 2 into a generic ScyllaDBTemplate which will make it uniform with the `rackTemplate` and picking particular overrides at the racks level

Reply:
i don't understand what you mean. Can you provide a code change?
Keep in mind that all properties of non-embedded part of ScyllaDB struct are global to datacenter and cannot be specified on rack level.

Reply:
I am proposing to be able to template+override the rest bellow, like `AdditionalScyllaDBArguments`

`AdditionalScyllaDBArguments` can differ by racks, if you want to test a limited change
`Image` can differ if you want to test it in a particular racks.
Similarly for the rest. All of them can technically  differ by the nature of a rolling update. Unifying the approach would make this consistent, even though some cases are not mainstream.

Reply:
ok, let's leave it like this for now. Since all have effectively a default value, we can talk individual points later.